File: src/structural/flyweight/Flyweight.cs
namespace DesignPatterns.Structural.Flyweight;

// Flyweight interface
public interface IFlyweight
{
    void Operation(string extrinsicState);
}

// Concrete Flyweight
public class ConcreteFlyweight : IFlyweight
{
    private string _intrinsicState;

    public ConcreteFlyweight(string intrinsicState)
    {
        _intrinsicState = intrinsicState;
    }

    public void Operation(string extrinsicState)
    {
        Console.WriteLine($"ConcreteFlyweight: Intrinsic State = {_intrinsicState}, Extrinsic State = {extrinsicState}");
    }
}

// Flyweight Factory
public class FlyweightFactory
{
    private Dictionary<string, IFlyweight> _flyweights = new Dictionary<string, IFlyweight>();

    public IFlyweight GetFlyweight(string key)
    {
        if (!_flyweights.ContainsKey(key))
        {
            _flyweights[key] = new ConcreteFlyweight(key);
        }

        return _flyweights[key];
    }
}

// Client code
class Program
{
    static void Main()
    {
        FlyweightFactory factory = new FlyweightFactory();

        var flyweight1 = factory.GetFlyweight("A");
        var flyweight2 = factory.GetFlyweight("B");
        var flyweight3 = factory.GetFlyweight("A");

        flyweight1.Operation("First Call");
        flyweight2.Operation("Second Call");
        flyweight3.Operation("Third Call");

        Console.WriteLine(object.ReferenceEquals(flyweight1, flyweight3)); // True
    }
}

File: structural/flyweight/flyweight.py

# Flyweight interface
class Flyweight:
    def operation(self, extrinsic_state):
        pass

# Concrete Flyweight
class ConcreteFlyweight(Flyweight):
    def __init__(self, intrinsic_state):
        self._intrinsic_state = intrinsic_state

    def operation(self, extrinsic_state):
        print(f"ConcreteFlyweight: Intrinsic State = {self._intrinsic_state}, Extrinsic State = {extrinsic_state}")

# Flyweight Factory
class FlyweightFactory:
    def __init__(self):
        self._flyweights = {}

    def get_flyweight(self, key):
        if key not in self._flyweights:
            self._flyweights[key] = ConcreteFlyweight(key)
        return self._flyweights[key]

if __name__ == "__main__":
    factory = FlyweightFactory()

    flyweight1 = factory.get_flyweight("A")
    flyweight2 = factory.get_flyweight("B")
    flyweight3 = factory.get_flyweight("A")

    flyweight1.operation("First Call")
    flyweight2.operation("Second Call")
    flyweight3.operation("Third Call")

    print(flyweight1 is flyweight3)  # True
